File: hackerrank/java/balancedbrackets/src/BalancedBrackets.java
import java.util.Scanner;
import java.util.Stack;

public class BalancedBrackets {
    private static boolean isBalanced(String expression) {
        Stack<Character> s = new Stack<>();
        for (int i = 0; i < expression.length(); i++) {
            char ch = expression.charAt(i);
            switch (ch) {
                case '{':
                    s.push(ch);
                    break;
                case '[':
                    s.push(ch);
                    break;
                case '(':
                    s.push(ch);
                    break;
                default:
                    if (s.empty() || !isPair(ch, s.pop())) return false;
            }
        }
        return s.empty();
    }

    private static boolean isPair(char a, char b){
        int asciiA = (int) a;
        //System.out.println(asciiA);
        int asciiB = (int) b;
        //System.out.println(asciiB);
        if((asciiA == 123 && asciiB == 125) || (asciiA == 125 && asciiB == 123))
            return true;
        else if((asciiA == 91 && asciiB == 93) || (asciiA == 93 && asciiB == 91))
            return true;
        else
            return (asciiA == 41 && asciiB == 40) || (asciiA == 40 && asciiB == 41);
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int t = in.nextInt();
        for (int a0 = 0; a0 < t; a0++) {
            String expression = in.next();
            System.out.println( (isBalanced(expression)) ? "YES" : "NO" );
        }
    }
}

File: hackerrank/python/balanced_brackets.py
#!usr/bin/env ipython

"""Balanced Brackets (stacks)
we say a sequence of brackets is considered to be balanced if the following 
conditions are met:
1. It contains no unmatched brackets.
2. The subset of brackets enclosed within the confines of a matched pair of 
brackets is also a matched pair of brackets.
Given  strings of brackets, determine whether each sequence of brackets 
is balanced. If a string is balanced, print YES on a new line; 
otherwise, print NO on a new line.
"""

# first, find the ascii code of six types of brackets
# second, write a matcher to pair brackets
# third, use stack to store the input string, pop from stack to create reverse string,
# fourth, for the input and reverse strings, check if ith position form a brackets
def is_matched(expression):
    """
    @param input: string
    @return bool
    """
    # use list as stack
    st = list()
    for i in expression:
        if i == '{' or i == '[' or i == '(':
            st.append(i)
        elif i == '}' or i == ']' or i == ')':
            if not st or is_pair(st.pop(), i) == False:
                return False
    if st:
        return False
    return True        

def is_pair(a, b):
    """
    @param a: char
    @param b: char
    @return bool
    """
    ascii_a = ord(a)
    ascii_b = ord(b)
    if ascii_a == 123 and ascii_b == 125: # { and }
        return True
    elif ascii_a == 125 and ascii_b == 123:
        return True
    elif ascii_a == 91 and ascii_b == 93: # [ and ]
        return True
    elif ascii_a == 93 and ascii_b == 91:
        return True
    elif ascii_a == 40 and ascii_b == 41: # ( and )
        return True
    elif ascii_a == 41 and ascii_b == 40:
        return True
    else:
        return False

if __name__ == '__main__':

    t = int(raw_input().strip())
    for a0 in xrange(t):
        expression = raw_input().strip()
        if is_matched(expression) == True:
            print "YES"
        else:
            print "NO"
